lib/kubernetes-deploy/deploy_task.rb

```diff
@@ -204,15 +200,17 @@ def cluster_resource_discoverer
       )
     end
 
-    def ejson_provisioner
-      @ejson_provisioner ||= EjsonSecretProvisioner.new(
-        namespace: @namespace,
-        context: @context,
-        template_dir: @template_dir,
-        logger: @logger,
-        statsd_tags: @namespace_tags,
-        selector: @selector,
-      )
+    def ejson_provisioners
+      @ejson_provisioners ||= TemplateDiscovery.templates(@template_paths).keys.map do |template_dir|
```

Comment:
This feels a bit messy, but perhaps this is something to look at in the future. We need to call `TemplateDiscovery` in order to get a nice aggregation of explicit file arguments into a single template dir. If we don't do that, we'd need to add additional logic in the secret provisioner as to whether it's dealing with a directory or a file.... that's what `TemplateDiscovery` is supposed to take care of!

Reply:
What happens if I pass `-f config/deploy/staging/` and `-f config/deploy/production/i_dont_like_sym_links.yml` and there are ejson files in each directory?

Reply:
Ah, you're right, that will deploy the undeclared secret :( I've got a (failing) test for this written now, just trying to decide how best to solve it.

It seems like we need to pass the full template mappings around... I'm not sure I want to call `@templates = TemplateDiscovery.new(@template_paths).templates` right in the constructor (is it doing too much to be there?). What are your thoughts. I recall Katrina mentioning that we should do as little as possible in the constructors. Said that, `TemplateDiscovery#templates` can't really throw any errors so it might be safe to do

Reply:
If we're trying not to break backwards compatibility adding a method like `TemplateDiscovery.ejson_templates(@template_paths).keys.map do ...` might be the simplest solution. If we we're ok breaking backwards compatibility I could see `EjsonProvisioner` taking a file instead of a directory in which its looking for a specific file.

Reply:
Let's not break backwards-compat just yet. I like your idea of `TemplateDiscovery.ejson_templates`